### PR TITLE
Support RDD comparisions with order and unequal partitioners

### PR DESCRIPTION
**Pull request by holdenk**

### CODE REVIEW

File: src/main/1.3/scala/com/holdenkarau/spark/testing/RDDComparisons.scala [hunk 1]
```diff
@@ -26,12 +26,26 @@ import org.apache.spark.rdd._
 
 
 object RDDComparisons {
+  /**
+   * Compare two RDDs. If they are equal returns None, otherwise
+   * returns Some with the first mismatch. Requires that expected has an explicit partitioner.
```
Comment (mahmoudhanafy) [inline src/main/1.3/scala/com/holdenkarau/spark/testing/RDDComparisons.scala]: What if expected doesn't have explicit partitioner ?

File: src/main/1.3/scala/com/holdenkarau/spark/testing/RDDComparisons.scala [hunk 2]
```diff
@@ -26,12 +26,26 @@ import org.apache.spark.rdd._
...
+   */
+  def compareWithOrder[T: ClassTag](expected: RDD[T], result: RDD[T]): Option[(T, T)] = {
+    if (result.partitioner.map(_ == expected.partitioner.get).getOrElse(false)) {
+      compareWithOrderSamePartitioner(expected, result)
+    } else {
+      val targetPartitions = expected.partitions.size
```
Comment (mahmoudhanafy) [inline src/main/1.3/scala/com/holdenkarau/spark/testing/RDDComparisons.scala]: Could you generate test case to test this part ?

Comment (holdenk) [inline src/main/1.3/scala/com/holdenkarau/spark/testing/RDDComparisons.scala]: Good idea, I'll add some tests for this.

File: src/main/1.3/scala/com/holdenkarau/spark/testing/RDDComparisons.scala [hunk 3]
```diff
@@ -26,12 +26,26 @@ import org.apache.spark.rdd._
...
+      compareWithOrderSamePartitioner(expected.repartition(targetPartitions),
```
Comment (mahmoudhanafy) [inline src/main/1.3/scala/com/holdenkarau/spark/testing/RDDComparisons.scala]: Why you repartition expected ?? It is already has this partitioner

File: src/main/1.3/scala/com/holdenkarau/spark/testing/RDDComparisons.scala [hunk 4]
```diff
@@ -26,13 +26,57 @@ import org.apache.spark.rdd._
...
+   */
+  def compareWithOrder[T: ClassTag](expected: RDD[T], result: RDD[T]): Option[(Option[T], Option[T])] = {
+    // If there is a known partitioner just zip
+    if (result.partitioner.map(_ == expected.partitioner.get).getOrElse(false)) {
+      compareWithOrderSamePartitioner(expected, result)
+    } else {
+      // Otherwise index every element
+      def indexRDD[T](rdd: RDD[T]): RDD[(Int, T)] = {
+        val counts = rdd.mapPartitions(itr => List(itr.size).toIterator).collect()
+        val indexed = rdd.mapPartitionsWithIndex{
```
Comment (mahmoudhanafy) [inline src/main/1.3/scala/com/holdenkarau/spark/testing/RDDComparisons.scala]: Can this be valid ?
```Scala
val indexed = rdd.zipWithIndex().map(_.swap)
```

File: src/main/1.3/scala/com/holdenkarau/spark/testing/RDDComparisons.scala [hunk 5]
```diff
@@ -26,13 +26,57 @@ import org.apache.spark.rdd._
...
+        def next(): (Option[T], Option[T]) = {
+          (thisIter.hasNext, otherIter.hasNext) match {
+            case (false, true) => (Option.empty[T], Some(otherIter.next()))
+            case (true, false) => (Some(thisIter.next()), Option.empty[T])
+            case (true, true) =>  (Some(thisIter.next()), Some(otherIter.next()))
+            case _ => throw new Exception("next called when elements consumed")
+          }
+        }
+      }
+    }.filter{case (v1, v2) => v1 == v2}.take(1).headOption
```
Comment (mahmoudhanafy) [inline src/main/1.3/scala/com/holdenkarau/spark/testing/RDDComparisons.scala]: Should it be `v1 != v2` ?

File: src/main/1.3/scala/com/holdenkarau/spark/testing/RDDComparisons.scala [hunk 6]
```diff
@@ -26,13 +26,57 @@ import org.apache.spark.rdd._
...
+    // If there is a known partitioner just zip
+    if (result.partitioner.map(_ == expected.partitioner.get).getOrElse(false)) {
+      compareWithOrderSamePartitioner(expected, result)
+    } else {
+      // Otherwise index every element
+      def indexRDD[T](rdd: RDD[T]): RDD[(Int, T)] = {
+        val counts = rdd.mapPartitions(itr => List(itr.size).toIterator).collect()
+        val indexed = rdd.mapPartitionsWithIndex{
+          case (idx, itr) =>
+            val start = if (idx == 0) 0 else counts(idx-1)
```
Comment (mahmoudhanafy) [inline src/main/1.3/scala/com/holdenkarau/spark/testing/RDDComparisons.scala]: I think here you should start from the sum of all previous partitions' count, not only from the last partition count.

File: src/main/1.3/scala/com/holdenkarau/spark/testing/RDDComparisons.scala (hunk 4 above)
Comment (holdenk) [inline src/main/1.3/scala/com/holdenkarau/spark/testing/RDDComparisons.scala]: Sadly zipwithindex isn't defined on RDDs.

File: src/main/1.3/scala/com/holdenkarau/spark/testing/RDDComparisons.scala (hunk 6 above)
Comment (holdenk) [inline src/main/1.3/scala/com/holdenkarau/spark/testing/RDDComparisons.scala]: Ah yes good catch. I'll switch our test to use 5 partitions so it catches this too.

File: src/main/1.3/scala/com/holdenkarau/spark/testing/RDDComparisons.scala (hunk 4 above)
Comment (mahmoudhanafy) [inline src/main/1.3/scala/com/holdenkarau/spark/testing/RDDComparisons.scala]: I found it [here](https://spark.apache.org/docs/1.3.0/api/java/org/apache/spark/rdd/RDD.html#zipWithIndex()) at the docs.

Comment (holdenk) [inline src/main/1.3/scala/com/holdenkarau/spark/testing/RDDComparisons.scala]: clearly I haven't been keeping up with that enough, thanks.

**Comment by mahmoudhanafy** [comment]
It looks good now.
What about creating method to compare PairRDD RDD[(K, V)] ? We can add a restriction that the keys should be unique. This could be also in another PR

**Pull request merged**